File: python/tests/runtime_fixture_commands.py
from __future__ import annotations

from collections.abc import Sequence

from onetool.commands import CommandSpec
from onetool.types import err, ok, ok_bytes


def runtime_fixture_commands(ids: Sequence[str]) -> tuple[CommandSpec, ...]:
    fixtures: list[CommandSpec] = []

    for command_id in ids:
        fixtures.append(_runtime_fixture_command(command_id))

    return tuple(fixtures)


def _runtime_fixture_command(command_id: str) -> CommandSpec:
    match command_id:
        case "echo":
            return CommandSpec(
                name="echo",
                summary="Echo text.",
                usage="echo <text...>",
                details="Examples:\n  echo hello",
                handler=lambda _ctx, args, _stdin: ok(" ".join(args)),
            )
        case "fail":
            return CommandSpec(
                name="fail",
                summary="Fail immediately.",
                usage="fail",
                details="Examples:\n  fail",
                handler=lambda _ctx, _args, _stdin: err("fail: forced failure"),
            )
        case "burst":
            return CommandSpec(
                name="burst",
                summary="Emit three lines.",
                usage="burst",
                details="Examples:\n  burst",
                handler=lambda _ctx, _args, _stdin: ok("first line\nsecond line\nthird line"),
            )
        case "binary":
            return CommandSpec(
                name="binary",
                summary="Emit binary bytes.",
                usage="binary",
                details="Examples:\n  binary",
                handler=lambda _ctx, _args, _stdin: ok_bytes(bytes([0, 1, 2, 3]), "application/octet-stream"),
            )
        case "longline":
            return CommandSpec(
                name="longline",
                summary="Emit one long line.",
                usage="longline",
                details="Examples:\n  longline",
                handler=lambda _ctx, _args, _stdin: ok("abcdefghijklmnopqrstuvwxyz"),
            )
        case "help-override":
            return CommandSpec(
                name="help",
                summary="Overridden help.",
                usage="help",
                details="Examples:\n  help",
                handler=lambda _ctx, _args, _stdin: ok("override help"),
            )
        case _:
            raise AssertionError(f"unsupported runtime fixture command: {command_id}")

File: test/runtime-fixture-commands.ts
import type { CommandSpec } from '../src/commands/index.js';
import { err, ok, okBytes, type CommandResult } from '../src/types.js';

export type RuntimeFixtureCommandId = 'echo' | 'fail' | 'burst' | 'binary' | 'longline' | 'help-override';

export function runtimeFixtureCommands(ids: readonly RuntimeFixtureCommandId[]): CommandSpec[] {
  return ids.map(function (id) {
    switch (id) {
      case 'echo':
        return {
          name: 'echo',
          summary: 'Echo text.',
          usage: 'echo <text...>',
          details: 'Examples:\n  echo hello',
          async handler(_ctx, args): Promise<CommandResult> {
            return ok(args.join(' '));
          },
        };
      case 'fail':
        return {
          name: 'fail',
          summary: 'Fail immediately.',
          usage: 'fail',
          details: 'Examples:\n  fail',
          async handler(): Promise<CommandResult> {
            return err('fail: forced failure');
          },
        };
      case 'burst':
        return {
          name: 'burst',
          summary: 'Emit three lines.',
          usage: 'burst',
          details: 'Examples:\n  burst',
          async handler(): Promise<CommandResult> {
            return ok('first line\nsecond line\nthird line');
          },
        };
      case 'binary':
        return {
          name: 'binary',
          summary: 'Emit binary bytes.',
          usage: 'binary',
          details: 'Examples:\n  binary',
          async handler(): Promise<CommandResult> {
            return okBytes(Uint8Array.of(0, 1, 2, 3), 'application/octet-stream');
          },
        };
      case 'longline':
        return {
          name: 'longline',
          summary: 'Emit one long line.',
          usage: 'longline',
          details: 'Examples:\n  longline',
          async handler(): Promise<CommandResult> {
            return ok('abcdefghijklmnopqrstuvwxyz');
          },
        };
      case 'help-override':
        return {
          name: 'help',
          summary: 'Overridden help.',
          usage: 'help',
          details: 'Examples:\n  help',
          async handler(): Promise<CommandResult> {
            return ok('override help');
          },
        };
      default: {
        const _never: never = id;
        return _never;
      }
    }
  });
}
